lib/client/doc.js

```diff
@@ -977,48 +973,49 @@ Doc.prototype._opAcknowledged = function(message) {
   this._clearInflightOp();
 };
 
-Doc.prototype._rollback = function(err) {
+Doc.prototype._rollback = function(error) {
   // The server has rejected submission of the current operation. Invert by
   // just the inflight op if possible. If not possible to invert, cancel all
   // pending ops and fetch the latest from the server to get us back into a
   // working state, then call back
   var op = this.inflightOp;
 
-  if ('op' in op && op.type.invert) {
-    try {
-      op.op = op.type.invert(op.op);
-    } catch (error) {
-      // If the op doesn't support `.invert()`, we just reload the doc
-      // instead of trying to locally revert it.
-      return this._hardRollback(err);
-    }
+  if (!('op' in op && op.type.invert)) {
+    return this._hardRollback(error);
+  }
 
-    // Transform the undo operation by any pending ops.
-    for (var i = 0; i < this.pendingOps.length; i++) {
-      var transformErr = transformX(this.pendingOps[i], op);
-      if (transformErr) return this._hardRollback(transformErr);
-    }
+  try {
+    op.op = op.type.invert(op.op);
+  } catch (invertOpError) {
+    logger.info('Cannot invert op', op, invertOpError);
+    // If the op doesn't support `.invert()`, we just reload the doc
+    // instead of trying to locally revert it.
+    return this._hardRollback(error);
+  }
 
-    // ... and apply it locally, reverting the changes.
-    //
-    // This operation is applied to look like it comes from a remote source.
-    // I'm still not 100% sure about this functionality, because its really a
-    // local op. Basically, the problem is that if the client's op is rejected
-    // by the server, the editor window should update to reflect the undo.
-    try {
-      this._otApply(op, false);
-    } catch (error) {
-      return this._hardRollback(error);
-    }
+  // Transform the undo operation by any pending ops.
+  for (var i = 0; i < this.pendingOps.length; i++) {
+    var transformErr = transformX(this.pendingOps[i], op);
+    if (transformErr) return this._hardRollback(transformErr);
+  }
 
-    this._clearInflightOp(err);
-    return;
+  // ... and apply it locally, reverting the changes.
+  //
+  // This operation is applied to look like it comes from a remote source.
+  // I'm still not 100% sure about this functionality, because its really a
+  // local op. Basically, the problem is that if the client's op is rejected
+  // by the server, the editor window should update to reflect the undo.
+  try {
+    this._otApply(op, false);
+  } catch (otApplyError) {
+    logger.info('Cannot apply op in rollback.', op, otApplyError, 'Original error', error);
+    return this._hardRollback(otApplyError);
   }
 
-  this._hardRollback(err);
+  this._clearInflightOp(error);
```

Comment:
We can probably keep the old `OP_SUBMIT_REJECTED` logic here, since at this point in the code we've successfully rolled back the inflight op with no further errors, so we can swallow the error as originally intended.

```suggestion
  // The server has rejected submission of the current operation. If we get
  // an "Op submit rejected" error, this was done intentionally
  // and we should roll back but not return an error to the user.
  if (error.code !== ERROR_CODE.ERR_OP_SUBMIT_REJECTED) {
    this._clearInflightOp(error);
  }
```